zeus/optimizer/batch_size/server/optimizer.py

```diff
@@ -0,0 +1,310 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from uuid import UUID
+from fastapi import HTTPException
+
+import numpy as np
+from zeus.optimizer import batch_size
+
+from zeus.optimizer.batch_size.server.models import (
+    JobSpec,
+    ReportResponse,
+    TrainingResult,
+)
+from zeus.optimizer.batch_size.server.mab import GaussianTS
+from zeus.optimizer.batch_size.server.explorer import PruningExploreManager
+from zeus.util.metric import zeus_cost
+
+# from dbapis import DBAPI
+
+GLOBAL_ZEUS_SERVER: ZeusBatchSizeOptimizer | None = None
+
+"""
+TODO: ADD EARLY STOPPING SOMEHOW?. If cost is too much, stop training 
+"""
+
+
+class ZeusBatchSizeOptimizer:
+    """A singleton class that manages training jobs. Why singleton? Not sure yet.
+    (1) Does this class maintain any state?
+    Or (2)are we storing everything in DB and everytime we call a function, it just grab values from DB and compute
+
+    Pruning stage will be handled by throwing an error such as cannot find batch size that converges.
+    User provides max Epoch, and if the cost is too much or cannot achieve target accuracy, we throw convergence fail.
+
+    So all we need to do is, In DB, we specify if we are in pruning stage(filterting batch sizes that cannot achieve target acc)
+    OR training stage(have some batch sizes that can achieve target accuracy)
+    """
+
+    def __init__(
+        self,
+        verbose: bool = True,
+    ) -> None:
+        """Initialize the server."""
+        self.concurrency = True
+        # self.seed = seed
+        self.verbose = verbose
+
+        # One MAB for each job.
+        self.mabs: dict[UUID, GaussianTS] = {}
+
+        # Observation history (batch size, reward) for each job.
+        self.history: dict[UUID, defaultdict[int, list[float]]] = {}
+
+        # Min_cost observed so far
+        self.min_costs: dict[UUID, float] = {}
+
+        # One PruningExplorationManager for each job.
+        self.exp_manager: dict[UUID, PruningExploreManager] = {}
+
+        # Job UUID -> Job spec (will be replaced to hit DB prob)
+        self.jobs: dict[UUID, JobSpec] = {}
+
+        # # Track the batch size range for each job.
+        # self.batch_sizes: dict[JobSpec, list[int]] = {}
+
+    @property
+    def name(self) -> str:
+        """Name of the batch size optimizer."""
+        return "Pruning GaussianTS BSO"
+
+    def register_job(self, job: JobSpec) -> int:
+        """Register a user-submitted job. Return number of newly created job. Return the number of job that is registered"""
+        if job.job_id in self.jobs:
+            if self.verbose:
+                self._log(f"Job({job.job_id}) already exists")
+            if self.jobs[job.job_id] != job:
+                raise HTTPException(
```

Comment:
This is a general design guideline: Only the router gets to raise HTTPExceptions and decide the HTTP response code. Then, you consolidate the handling of **all possible exceptions that can be raised** in the handler function in `router.py`. Specifically, here, you define `ZeusBSOJobSpecMismatch` under something like `zeus.optimizer.batch_size.models/common` and raise that. Then, the handler function has a gigantic `try`-`except` clause over the call of each `ZeusBatchSizeOptimizer`'s method with one `except` clause for every possible exception that can possibly be raised by calling that method, and raises the appropriate `HTTPException` with an appropriate status code.

Reply:
https://fastapi.tiangolo.com/tutorial/handling-errors/#:~:text=%3D%20FastAPI()-,%40app.exception_handler(UnicornException),),-%40app.get

Something like this?

Reply:
Oh interesting, it allows one to wrap the try-except clause with an exception handler. But I would say writing exception handlers could be an overkill as they are mostly intended for reuse when you have many many endpoints calling an overlapping set of methods. I won't stop you from doing it because exception handlers are nicer, but I think I would probably just use a dirty try-except-except-except-... statement for now.